pkg/acpi/coreboot/madt.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package coreboot
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/u-root/u-root/pkg/acpi"
+)
+
+const (
+	minMADT = 8 // 32-bit address and flags
+)
+const (
+	entryLAPIC         = 0
+	entryIOAPIC        = 1
+	entryISOR          = 2 // pronounced: eyesore, for reasons.
+	entryBOGUS         = 3
+	entryNMI           = 4
+	entryLAPICOverRide = 5
+	entryLX2APIC       = 9
+	entryLX2NMI        = 10
+
+	sizeEntryLAPIC         = 8
+	sizeEntryIOAPIC        = 12
+	sizeEntryISOR          = 10 // pronounced: eyesore, for reasons.
+	sizeEntryNMI           = 6
+	sizeEntryLAPICOverRide = 12
+	sizeEntryLX2APIC       = 16
+	sizeEntryLX2NMI        = 12
+)
+
+var (
+	// The MADT signature is APIC.
+	// Another failure of vision; they never thought there'd
+	// be more than one APIC I guess.
+	sigMADT = "APIC"
+	info    = map[int]struct {
+		name string
+		size int
+	}{
+		entryLAPIC:         {"LAPIC", sizeEntryLAPIC},
+		entryIOAPIC:        {"IOAPIC", sizeEntryIOAPIC},
+		entryISOR:          {"ISOR", sizeEntryISOR},
+		entryNMI:           {"NMI", sizeEntryNMI},
+		entryLAPICOverRide: {"LAPICOverRide", sizeEntryLAPICOverRide},
+		entryLX2APIC:       {"LX2APIC", sizeEntryLX2APIC},
+		entryLX2NMI:        {"X2NMI", sizeEntryLX2NMI},
+	}
+)
+
+// MADT defines the madt struct.
+type MADT struct {
+	acpi.Table
+	// base is the base address of the MADT struct in physical memory.
+	base int64
+}
+
+func init() {
+	corebooters["APIC"] = NewMADT
+}
+
+// rv reverses a slice
+func rv(b []byte) []byte {
+	for i := 0; i < len(b)/2; i++ {
+		j := len(b) - i - 1
+		b[i], b[j] = b[j], b[i]
+	}
+	return b
+}
+
+// NewMADT returns a MADT, from a Table, iff the signature is correct.
+func NewMADT(m acpi.Table) (Corebooter, error) {
+	if m.Sig() != sigMADT {
+		return nil, fmt.Errorf("%s is not a MADT", m.Sig())
+	}
+	b := m.TableData()
+	Debug("NEWMADT Subtables: %d bytes", len(b))
+	if len(b) < minMADT {
+		return nil, fmt.Errorf("Table is only %d bytes, need %d", len(m.TableData()), minMADT)
+	}
+	b = b[8:]
+	Debug("NEWMADT Subtables: after removing 8 bytes of fluff, %d bytes", len(b))
+	for len(b) > 0 {
```

Comment:
can you describe in a comment the MADT format?

// MADT consists of TLV entries: 1-byte type, 1-byte length, value.